Simulator/Simulator/Simulator/python/utils/utils_matplotlib.py

```diff
@@ -60,26 +195,93 @@ def line_plot(output_file, y_data, line_label, x_label, y_label, title):
+            y_min, y_max = get_min_max(y_min, y_max, y_data[i])
 
-        x_min = x_data_min if x_data_min < x_min else x_min
-        x_max = x_data_max if x_data_max > x_max else x_max
-        y_min = y_data_min if y_data_min < y_min else y_min
-        y_max = y_data_max if y_data_max > y_max else y_max
+    set_x_ticks(x_min, x_max)
+    set_y_ticks(y_min, y_max)
+    set_legend()
 
-    x_step = (x_max - x_min) / 40.0
-    if x_step != 0.0:
-        plt.xticks(np.arange(start=x_min, stop=x_max + x_step, step=x_step), rotation=90)
+    plt.grid(True, linestyle='--')
+    plt.title(title, fontsize=20)
+    plt.xlabel(x_label, fontsize=20)
+    plt.ylabel(y_label, fontsize=20)
+    plt.tight_layout()
 
-    y_step = (y_max - y_min) / 30.0
-    if y_step != 0.0:
-        plt.yticks(np.arange(start=y_min, stop=y_max + y_step, step=y_step))
+    plt.savefig(output_file, dpi=100)
+
+    reset_plot()
+
+
+def order_points_to_polygon(x_data, y_data):
+    '''
+    It orders a set of points so that they form a polygon
+    https://stackoverflow.com/questions/10846431/ordering-shuffled-points-that-can-be-joined-to-form-a-polygon-in-python
+    '''
+    pp = [(x, y) for x, y in zip(x_data, y_data)]
+    # sort by polar angle
+    cent = (sum([p[0] for p in pp]) / len(pp), sum([p[1] for p in pp]) / len(pp))
+    pp.sort(key=lambda p: math.atan2(p[1] - cent[1], p[0] - cent[0]))
+    return np.array([p[0] for p in pp]), np.array([p[1] for p in pp])
+
+
+def fill_plot(output_file, groups, title, x_label, y_label):
+    """
+    This will plot multiple cluster/polygons/groups of points, each with a different color
+
+    :param output_file: str
+        path to the file where to save the plot
+
+    :param groups: dict
+        each key represents the label for one group of points
+        each values is a dictionary like:
+        {
+            'x': x coordinates for the x axis
+            'y': y coordinates for the y axis
+        }
+
+    :param title: str
+        title of the plot
+
+    :param x_label: str
+        label for the x axis
+
+    :param y_label: str
+        label for the y axis
+    """
+    # groupts is a dict
+
+    plt.figure(figsize=(15, 10))
+
+    # duplicate code
+    x_min = 1e+15
+    x_max = -1e+15
+    y_min = 1e+15
+    y_max = -1e+15
+
+    for group_name in groups.keys():
+        group_data = groups[group_name]
+
+        x_data = group_data['x']
+        y_data = group_data['y']
+
+        if x_data.shape[0] == 0:
+            continue
+
+        # create polygon from points
+        #x_data, y_data = order_points_to_polygon(x_data, y_data)
+
+        #plt.fill(x_data, y_data, label=group_name)
+        plt.scatter(x_data, y_data, label=group_name)
+        #plt.text(np.mean(x_data), np.mean(y_data), group_name)
+        x_min, x_max = get_min_max(x_min, x_max, x_data)
+        y_min, y_max = get_min_max(y_min, y_max, y_data)
+
+    set_x_ticks(x_min, x_max)
+    set_y_ticks(y_min, y_max)
+    set_legend()
```

Comment:
Update this function, we will not use the fill anymore, its to hard to make it work well

Reply:
This uses the simple ```scatter``` function